Zig CC: MacOS Linker: Symbol isn't Weak (`__attribute__((weak))` unsupported?)

### Zig Version

0.11.0

### Steps to Reproduce and Observed Behavior

The `__darwin_check_fd_set_overflow` function was introduced in macOS 11.0. It's undefined in older versions. The way macOS standard libraries deal with this is by making the symbol weak, and then checking if it's 0 or not at runtime.

The `__API_AVAILABLE` macro should take care of making the symbol weak:

/usr/include/sys/_types/_fd_def.h:
```c
int __darwin_check_fd_set_overflow(int, const void *, int) __API_AVAILABLE(macosx(11.0), ios(14.0), tvos(14.0), watchos(7.0));   
```
When I build a binary targeting macOS 10.13 with Zig, I expect the symbol to be weakly linked. Unfortunately, that is not the case. When I build the same program with Apple-provided Clang, the symbol is weakly linked.

### Steps to Reproduce

1. Create a simple program using the `__darwin_check_fd_set_overflow` symbol:

test.c:
```c
#include <stdio.h>
#include <sys/_types/_fd_def.h>

int main(int argc, char** argv) {
    printf("__darwin_check_fd_set_overflow: %p\n", __darwin_check_fd_set_overflow);
}
```

2. Build it with zig cc:
```shell
$  zig cc --target=x86_64-macos.10.13 -isysroot $macos_sdk -iwithsysroot /usr/include --sysroot $macos_sdk -L/usr/lib -mmacos-version-min=10.13 test.c
```

3. Check if the symbol is weakly linked:

```shell
$ nm -m a.out
                 (undefined) external ___darwin_check_fd_set_overflow (from libSystem)
0000000100000000 (__TEXT,__text) [referenced dynamically] external __mh_execute_header
0000000100000580 (__TEXT,__text) external _main
                 (undefined) external _printf (from libSystem)
                 (undefined) external dyld_stub_binder (from libSystem)
```

the symbol is **not** weakly linked.

4. (optional) run the resulting binary on a 10.13 machine:
```shell
$ ./a.out
dyld: Symbol not found: ___darwin_check_fd_set_overflow
  Referenced from: //./a.out
  Expected in: /usr/lib/libSystem.B.dylib
 in //./a.out
```

Because the symbol is strongly linked, it inevitably fails.

### Expected Behavior

I expect zig cc's behavior to match Apple Clang's.

1. Build the same program with Clang on a Mac:
```shell
$ clang -mmacos-version-min=10.13 -arch x86_64 test.c                
test.c:5:52: warning: '__darwin_check_fd_set_overflow' is only available on macOS 11.0 or newer [-Wunguarded-availability-new]
    printf("__darwin_check_fd_set_overflow: %p\n", __darwin_check_fd_set_overflow);
                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_types/_fd_def.h:54:5: note: '__darwin_check_fd_set_overflow' has been marked as being introduced in macOS 11.0 here, but the deployment target is macOS 10.13.0
int __darwin_check_fd_set_overflow(int, const void *, int) __API_AVAILABLE(macosx(11.0), ios(14.0), tvos(14.0), watchos(7.0));
    ^
test.c:5:52: note: enclose '__darwin_check_fd_set_overflow' in a __builtin_available check to silence this warning
    printf("__darwin_check_fd_set_overflow: %p\n", __darwin_check_fd_set_overflow);
                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

2. Check if the symbol is weakly linked with nm:
```shell
$ nm -m a.out                                                                                                                                                                                                                                     
                 (undefined) weak external ___darwin_check_fd_set_overflow (from libSystem)
0000000100000000 (__TEXT,__text) [referenced dynamically] external __mh_execute_header
0000000100003f40 (__TEXT,__text) external _main
                 (undefined) external _printf (from libSystem)
                 (undefined) external dyld_stub_binder (from libSystem)
```
The symbol **is** weakly linked.

3. (optional) run the resulting binary on a 10.13 machine:
```shell
$ ./a.out 
__darwin_check_fd_set_overflow: 0x0
```
Things work as they should. The weakly linked symbol resolves to 0 if it's not found.


### Workarounds

So far the only workaround I've been able to find is patching Zig to link to weakly link `libSystem`, with a patch like this one:

```patch
diff --git a/src/link/MachO.zig b/src/link/MachO.zig
index 1f2dd4f8b..1ebd291ee 100644
--- a/src/link/MachO.zig
+++ b/src/link/MachO.zig
@@ -846,7 +846,7 @@ pub fn resolveLibSystem(
     });
     try out_libs.put(full_path, .{
         .needed = true,
-        .weak = false,
+        .weak = true,
         .path = full_path,
     });
 }
@@ -856,7 +856,7 @@ fn resolveLibSystemInDirs(arena: Allocator, dirs: []const []const u8, out_libs:
     // re-exports every single symbol definition.
     for (dirs) |dir| {
         if (try resolveLib(arena, dir, "System", ".tbd")) |full_path| {
-            try out_libs.put(full_path, .{ .needed = true, .weak = false, .path = full_path });
+            try out_libs.put(full_path, .{ .needed = true, .weak = true, .path = full_path });
             return true;
         }
     }
@@ -865,8 +865,8 @@ fn resolveLibSystemInDirs(arena: Allocator, dirs: []const []const u8, out_libs:
     for (dirs) |dir| {
         if (try resolveLib(arena, dir, "System", ".dylib")) |libsystem_path| {
             if (try resolveLib(arena, dir, "c", ".dylib")) |libc_path| {
-                try out_libs.put(libsystem_path, .{ .needed = true, .weak = false, .path = libsystem_path });
-                try out_libs.put(libc_path, .{ .needed = true, .weak = false, .path = libc_path });
+                try out_libs.put(libsystem_path, .{ .needed = true, .weak = true, .path = libsystem_path });
+                try out_libs.put(libc_path, .{ .needed = true, .weak = true, .path = libc_path });
                 return true;
             }
         }
```

This is not ideal, because I only want that one symbol to be weakly linked, not the entire library.

Additionally, this requires patching Zig, because unlike with Apple's Clang, `-weak-lSystem` cannot be passed in to make Zig link libSystem weakly because it hardcodes strong linking. (I filed another issue regarding this: https://github.com/ziglang/zig/issues/17149)

--- a/src/link/MachO/Dylib.zig
+++ b/src/link/MachO/Dylib.zig
@@ -520,7 +520,6 @@ pub fn resolveSymbols(self: *Dylib, macho_file: *MachO) void {
             global.nlist_idx = 0;
             global.file = self.index;
             global.flags.weak = flags.weak;
-            global.flags.weak_ref = false;
             global.flags.tlv = flags.tlv;
             global.flags.dyn_ref = false;
             global.flags.tentative = false;
@@ -534,9 +533,11 @@ pub fn resetGlobals(self: *Dylib, macho_file: *MachO) void {
         const sym = macho_file.getSymbol(sym_index);
         const name = sym.name;
         const global = sym.flags.global;
+        const weak_ref = sym.flags.weak_ref;
         sym.* = .{};
         sym.name = name;
         sym.flags.global = global;
+        sym.flags.weak_ref = weak_ref;
     }
 }
 

--- a/src/link/MachO/Object.zig
+++ b/src/link/MachO/Object.zig
@@ -525,6 +525,9 @@ fn initSymbols(self: *Object, macho_file: *MachO) !void {
             const off = try macho_file.strings.insert(gpa, name);
             const gop = try macho_file.getOrCreateGlobal(off);
             self.symbols.addOneAssumeCapacity().* = gop.index;
+            if (nlist.undf() and nlist.weakRef()) {
+                macho_file.getSymbol(gop.index).flags.weak_ref = true;
+            }
             continue;
         }
 
@@ -1099,9 +1102,11 @@ pub fn resetGlobals(self: *Object, macho_file: *MachO) void {
         const sym = macho_file.getSymbol(sym_index);
         const name = sym.name;
         const global = sym.flags.global;
+        const weak_ref = sym.flags.weak_ref;
         sym.* = .{};
         sym.name = name;
         sym.flags.global = global;
+        sym.flags.weak_ref = weak_ref;
     }
 }
 

--- a/src/link/MachO/ZigObject.zig
+++ b/src/link/MachO/ZigObject.zig
@@ -233,9 +233,11 @@ pub fn resetGlobals(self: *ZigObject, macho_file: *MachO) void {
         const sym = macho_file.getSymbol(sym_index);
         const name = sym.name;
         const global = sym.flags.global;
+        const weak_ref = sym.flags.weak_ref;
         sym.* = .{};
         sym.name = name;
         sym.flags.global = global;
+        sym.flags.weak_ref = weak_ref;
     }
 }
 

--- a/test/link/macho.zig
+++ b/test/link/macho.zig
@@ -43,6 +43,11 @@ pub fn testAll(b: *Build, build_opts: BuildOptions) *Step {
     macho_step.dependOn(testUnwindInfoNoSubsectionsX64(b, .{ .target = x86_64_target }));
     macho_step.dependOn(testUnwindInfoNoSubsectionsArm64(b, .{ .target = aarch64_target }));
     macho_step.dependOn(testWeakBind(b, .{ .target = x86_64_target }));
+    macho_step.dependOn(testWeakRef(b, .{ .target = b.resolveTargetQuery(.{
+        .cpu_arch = .x86_64,
+        .os_tag = .macos,
+        .os_version_min = .{ .semver = .{ .major = 10, .minor = 13, .patch = 0 } },
+    }) }));
 
     // Tests requiring symlinks when tested on Windows
     if (build_opts.has_symlinks_windows) {
@@ -2219,6 +2224,25 @@ fn testWeakLibrary(b: *Build, opts: Options) *Step {
     const run = addRunArtifact(exe);
     run.expectStdOutEqual("42 42");
     test_step.dependOn(&run.step);
+
+    return test_step;
+}
+
+fn testWeakRef(b: *Build, opts: Options) *Step {
+    const test_step = addTestStep(b, "macho-weak-ref", opts);
+
+    const exe = addExecutable(b, opts, .{ .name = "main", .c_source_bytes = 
+    \\#include <stdio.h>
+    \\#include <sys/_types/_fd_def.h>
+    \\int main(int argc, char** argv) {
+    \\    printf("__darwin_check_fd_set_overflow: %p\n", __darwin_check_fd_set_overflow);
+    \\}
+    });
+
+    const check = exe.checkObject();
+    check.checkInSymtab();
+    check.checkExact("(undefined) weakref external ___darwin_check_fd_set_overflow (from libSystem.B)");
+    test_step.dependOn(&check.step);
 
     return test_step;
 }

